### PR TITLE
Security: protect against invalid export dates

### PR DESCRIPTION
**Pull request by ravinderk**
Resolves #

## Description

While exporting revenue and donation stats, the start year and end year should be in the valid year range. The year value should be in the first donation and last donation year range. I update the business logic to revoke exporting process if the year is invalid.

#### Highlight
- throw an exception if any date param missing from request
- throw an exception upon an invalid year. it should be in the valid year range. the first year of donation to last year of donation is a valid donation year range.
-  show valid year choices in the year dropdown.

## Affects

It will affect `Export Revenue and Donation Stats` feature.

## Testing Instructions

Exported data should remain the same when compared with the latest GivewP release.

## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit and/or end-to-end tests
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
Thanks, @ravinderk! I tried exporting the donors with an empty start date and 104108731513857 as the end date... and it worked? I mean, it downloaded the donors and didn't crash anything. Does it just silently override invalid values?

File: includes/admin/tools/views/html-admin-page-exports.php [hunk 1]
```diff
@@ -77,14 +79,29 @@
...
-									/* translators: 1: start date dropdown 2: end date dropdown */
-									esc_html__( '%1$s to %2$s', 'give' ),
-									Give()->html->year_dropdown( 'start_year' ) . ' ' . Give()->html->month_dropdown( 'start_month' ),
-									Give()->html->year_dropdown( 'end_year' ) . ' ' . Give()->html->month_dropdown( 'end_month' )
-								);
-								?>
+                                <?php
+                                // @unreleased
+                                // Year in year downdown should begin from first donation year instead of only display first five recent year.
+                                $firstDonationDate = give(DonationRepository::class)->getFirstDonationDate();
```
[the remark this answers is not in the document]
Comment (ravinderk) [reply includes/admin/tools/views/html-admin-page-exports.php]: fixed: https://github.com/impress-org/givewp/compare/efc9efd746c6...74824a20313f

File: includes/admin/tools/export/class-export-earnings.php [hunk 2]
```diff
@@ -5,132 +5,172 @@
  * This class handles earnings export
```
[the remark this answers is not in the document]
Comment (ravinderk) [reply includes/admin/tools/export/class-export-earnings.php:5]: I made changes in the `get_data` function. Move the get date logic from request to  `getDatesFromRequest` function

**Comment by ravinderk** [comment]
> Thanks, @ravinderk! I tried exporting the donors with an empty start date and 104108731513857 as the end date... and it worked? I mean, it downloaded the donors and didn't crash anything. Does it just silently override invalid values?

@JasonTheAdams As per my understanding issue is with `Export Revenue and Donation Stats` instead of `Export Donors`. Let me know what do you think?

**Review by JasonTheAdams — APPROVED** [review]

**Comment by JasonTheAdams** [comment]
@ravinderk I changed this up a little bit. Instead of having the repository return `DateTime` objects, I had it return the model. Models are a great way to simplify a lot of logic with the trade-off that often more data is retrieved than is necessary (as in this case). But this way we avoid a bunch of incredibly specific, single purpose functions being added in places like the repository. Now we have two functions, one for grabbing the first and another the latest donation, that can be used in many different scenarios.

**Comment by ravinderk** [comment]
> @ravinderk I changed this up a little bit. Instead of having the repository return `DateTime` objects, I had it return the model. Models are a great way to simplify a lot of logic with the trade-off that often more data is retrieved than is necessary (as in this case). But this way we avoid a bunch of incredibly specific, single purpose functions being added in places like the repository. Now we have two functions, one for grabbing the first and another the latest donation, that can be used in many different scenarios.

@JasonTheAdams this make sense.

**Pull request merged**